cluster-autoscaler/cloudprovider/openstack/openstack_cloud_provider.go

```diff
@@ -0,0 +1,177 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openstack
+
+import (
+	"io"
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider"
+	"k8s.io/autoscaler/cluster-autoscaler/config"
+	"k8s.io/autoscaler/cluster-autoscaler/config/dynamic"
+	"k8s.io/autoscaler/cluster-autoscaler/utils/errors"
+	"k8s.io/klog"
+	"os"
+	"sync"
+)
+
+const (
+	// ProviderName is the cloud provider name for OpenStack
+	ProviderName = "openstack"
+)
+
+// OpenstackNodeGroup implements CloudProvider interface from cluster-autoscaler/cloudprovider module.
+type openstackCloudProvider struct {
+	openstackManager *OpenstackManager
+	resourceLimiter  *cloudprovider.ResourceLimiter
+	nodeGroups       []OpenstackNodeGroup
+}
+
+func buildOpenstackCloudProvider(openstackManager OpenstackManager, resourceLimiter *cloudprovider.ResourceLimiter) (cloudprovider.CloudProvider, error) {
+	os := &openstackCloudProvider{
+		openstackManager: &openstackManager,
+		resourceLimiter:  resourceLimiter,
+		nodeGroups:       []OpenstackNodeGroup{},
+	}
+	return os, nil
+}
+
+// Name returns the name of the cloud provider.
+func (os *openstackCloudProvider) Name() string {
+	return ProviderName
+}
+
+// NodeGroups returns all node groups managed by this cloud provider.
+func (os *openstackCloudProvider) NodeGroups() []cloudprovider.NodeGroup {
+	groups := make([]cloudprovider.NodeGroup, len(os.nodeGroups))
+	for i, group := range os.nodeGroups {
+		groups[i] = &group
+	}
+	return groups
+}
+
+// AddNodeGroup appends a node group to the list of node groups managed by this cloud provider.
+func (os *openstackCloudProvider) AddNodeGroup(group OpenstackNodeGroup) {
+	os.nodeGroups = append(os.nodeGroups, group)
+}
+
+// NodeGroupForNode returns the node group that a given node belongs to.
+//
+// Since only a single node group is currently supported, the first node group is always returned.
+func (os *openstackCloudProvider) NodeGroupForNode(node *apiv1.Node) (cloudprovider.NodeGroup, error) {
+	// TODO: wait for magnum nodegroup support
```

Comment:
Is there a storyboard entry we can link here?

Reply:
I've added a link where the node groups are created, to https://review.openstack.org/#/c/629274/
in https://github.com/kubernetes/autoscaler/pull/1690/commits/09e464a81edfadc84288c2237f7750528d431eef.